back-nest/src/products/products.services.ts

```diff
@@ -43,32 +103,56 @@ export class ProductService {
   }
 
   async upsertProducts(products: ScrapedProduct[]): Promise<void> {
-    await Promise.all(
+    await this.prisma.$transaction(
       products.map((productData) => {
-        const { title, source, price, newPrice, ...rest } = productData;
+        const { title, source, price, newPrice, profileImages, ...rest } =
+          productData;
+
+        const profileImagesArray: string[] = Array.isArray(profileImages)
+          ? profileImages
+          : profileImages
+            ? [profileImages]
+            : [];
 
         const productToUpdate = {
           ...rest,
           title,
           price,
           newPrice,
           source,
+          profileImages: profileImagesArray,
         };
 
         this.logger.log(`Upserting product: ${title} from source: ${source}`);
 
         return this.prisma.product.upsert({
           where: {
-            title_source: {
-              title,
-              source,
-            },
+            title_source: { title, source },
           },
           update: { ...productToUpdate },
           create: { ...productToUpdate },
         });
       }),
     );
+
+    const savedProducts = await this.prisma.product.findMany({
+      where: {
+        title: { in: products.map((p) => p.title) },
+      },
+    });
+
+    if (!savedProducts.length) {
+      this.logger.warn('No products were found in DB after upsert.');
+      return;
+    }
+
+    for (const product of savedProducts) {
+      await this.productSearchService.indexProduct(product);
+    }
```

Comment:
Check if you can speed things up here by using `Promise.all`

Reply:
Also, im highly recommending you to read about differences between [async loop iterations](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of) and [promise.all](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all)